src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,178 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.BraceCompletion
+{
+    internal abstract class AbstractBraceCompletionService : IBraceCompletionService
+    {
+        protected abstract char OpeningBrace { get; }
+
+        protected abstract char ClosingBrace { get; }
+
+        /// <summary>
+        /// Returns if the token is a valid opening token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidOpeningBraceToken(SyntaxToken token);
+
+        /// <summary>
+        /// Returns if the token is a valid closing token kind for this brace completion service.
+        /// </summary>
+        protected abstract bool IsValidClosingBraceToken(SyntaxToken token);
+
+        ///<inheritdoc cref="IBraceCompletionService.AllowOverTypeAsync(BraceCompletionContext, CancellationToken)"/>
+        public abstract Task<bool> AllowOverTypeAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken);
+
+        ///<inheritdoc cref="IBraceCompletionService.GetBraceCompletionAsync(BraceCompletionContext, CancellationToken)"/>
+        public async Task<BraceCompletionResult?> GetBraceCompletionAsync(BraceCompletionContext braceCompletionContext, CancellationToken cancellationToken)
+        {
+            var closingPoint = braceCompletionContext.ClosingPoint;
+            if (closingPoint < 1)
+            {
+                return null;
+            }
+
+            var openingPoint = braceCompletionContext.OpeningPoint;
+            var document = braceCompletionContext.Document;
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            if (sourceText[openingPoint] != OpeningBrace)
+            {
+                return null;
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(openingPoint, findInsideTrivia: true);
+            var validOpeningPoint = await IsValidOpenBraceTokenAtPositionAsync(token, openingPoint, document, cancellationToken).ConfigureAwait(false);
+            if (!validOpeningPoint)
+            {
+                return null;
+            }
+
+            var braceTextEdit = new TextChange(TextSpan.FromBounds(closingPoint, closingPoint), ClosingBrace.ToString());
+            // The caret location should be right before where the closing brace was inserted.
+            return new BraceCompletionResult(ImmutableArray.Create(braceTextEdit), caretLocation: openingPoint + 1);
```

Comment:
codedoesn't match comment.  comment talks aobut closing-brace, but code sets the caret based on open brace.